File: debop4s-data/src/test/scala/com/github/debop4s/data/tests/mapping/property/LocaledEntityTest.scala
package com.github.debop4s.data.tests.mapping.property

import com.github.debop4s.core.ValueObject
import com.github.debop4s.core.utils.Hashs
import com.github.debop4s.data.jpa.repository.JpaDao
import com.github.debop4s.data.model.{LongEntity, HibernateLocaleEntity, LocaleValue}
import com.github.debop4s.data.tests.AbstractJpaTest
import java.util
import java.util.Locale
import javax.persistence.Entity
import javax.persistence._
import org.hibernate.annotations._
import org.hibernate.{annotations => ha}
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.transaction.annotation.Transactional

/**
 * Created by debop on 2014. 3. 9.
 */
@Transactional
class LocaledEntityTest extends AbstractJpaTest {

  @Autowired val dao: JpaDao = null

  @Test
  def localedEntity() {
    val entity = new SampleLocaleEntity()
    entity.title = "제목-Original"
    entity.description = "설명-Original"
    entity.addLocaleValue(Locale.KOREAN, new SampleLocaleValue("제목", "설명"))
    entity.addLocaleValue(Locale.ENGLISH, new SampleLocaleValue("Title", "Description"))

    dao.persist(entity)
    dao.flush()
    dao.clear()

    val loaded = dao.findOne(classOf[SampleLocaleEntity], entity.id)
    assert(loaded != null)
    assert(loaded.localeMap.size() == 2)

    assert(loaded.getLocaleValue(Locale.KOREAN).title == "제목")
    assert(loaded.getLocaleValue(Locale.KOREAN).description == "설명")
    assert(loaded.getLocaleValue(Locale.ENGLISH).title == "Title")
    assert(loaded.getLocaleValue(Locale.ENGLISH).description == "Description")

    // 지원하지 않는 Locale 정보가 있다면, entity의 기본값을 사용합니다.
    assert(loaded.getLocaleValue(Locale.CHINESE).title == "제목-Original")
    assert(loaded.getLocaleValue(Locale.CHINESE).description == "설명-Original")

    dao.delete(loaded)
    dao.flush()

    assert(dao.findOne(classOf[SampleLocaleEntity], entity.id) == null)
  }

}

@Embeddable
class SampleLocaleValue extends ValueObject with LocaleValue {

  def this(title: String, description: String) {
    this()
    this.title = title
    this.description = description
  }

  var title: String = _
  var description: String = _

  override def hashCode(): Int = Hashs.compute(title)
}

@Entity
@DynamicInsert
@DynamicUpdate
class SampleLocaleEntity extends LongEntity with HibernateLocaleEntity[SampleLocaleValue] {

  var title: String = _
  var description: String = _

  @CollectionTable(name = "SampleLocaleEntityLocale", joinColumns = Array(new JoinColumn(name = "entityId")))
  @MapKeyClass(classOf[Locale])
  @ElementCollection(targetClass = classOf[SampleLocaleValue], fetch = FetchType.EAGER)
  @ha.Cascade(Array(ha.CascadeType.ALL))
  @LazyCollection(LazyCollectionOption.EXTRA)
  override val localeMap: util.Map[Locale, SampleLocaleValue] = new util.HashMap[Locale, SampleLocaleValue]()

  override protected def createDefaultLocaleVal: SampleLocaleValue = {
    new SampleLocaleValue(title, description)
  }

  @inline
  override def hashCode(): Int = Hashs.compute(title)
}

File: debop4s-data-orm/src/test/java/debop4s/data/orm/jtests/mapping/property/localed/LocaledEntityTest.java
package debop4s.data.orm.jtests.mapping.property.localed;

import debop4s.data.orm.jtests.jpa.JpaTestBase;
import lombok.extern.slf4j.Slf4j;
import org.junit.Test;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import java.util.Locale;

import static org.fest.assertions.Assertions.assertThat;

/**
 * debop4s.data.orm.s.mapping.property.localed.LocaledEntityTest
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 7. 22. 오후 5:58
 */
@Slf4j
@Transactional
public class LocaledEntityTest extends JpaTestBase {

    @PersistenceContext
    EntityManager em;

    @Test
    public void localedEntity() {
        SampleLocaleEntity entity = new SampleLocaleEntity();
        entity.setTitle("제목");
        entity.setDescription("내용");
        entity.addLocaleValue(Locale.KOREAN, new SampleLocaleValue("제목", "내용"));
        entity.addLocaleValue(Locale.ENGLISH, new SampleLocaleValue("Title", "Body"));

        em.persist(entity);
        em.flush();
        em.clear();

        SampleLocaleEntity loaded = em.find(SampleLocaleEntity.class, entity.getId());
        assertThat(loaded.getLocaleMap().size()).isEqualTo(2);

        assertThat(loaded.getLocaleValue(Locale.KOREAN).getTitle()).isEqualTo("제목");
        assertThat(loaded.getLocaleValue(Locale.KOREAN).getDescription()).isEqualTo("내용");
        assertThat(loaded.getLocaleValue(Locale.ENGLISH).getTitle()).isEqualTo("Title");
        assertThat(loaded.getLocaleValue(Locale.ENGLISH).getDescription()).isEqualTo("Body");

        // 없는 Locale 에 대해서는 기본 Locale 정보를 제공합니다.
        assertThat(loaded.getLocaleValue(Locale.CHINESE).getTitle()).isEqualTo("제목");
        assertThat(loaded.getLocaleValue(Locale.CHINESE).getDescription()).isEqualTo("내용");

        // second cache test
        loaded = em.find(SampleLocaleEntity.class, entity.getId());
        assertThat(loaded.getLocaleMap().size()).isEqualTo(2);

        assertThat(loaded.getLocaleValue(Locale.KOREAN).getTitle()).isEqualTo("제목");
        assertThat(loaded.getLocaleValue(Locale.KOREAN).getDescription()).isEqualTo("내용");
        assertThat(loaded.getLocaleValue(Locale.ENGLISH).getTitle()).isEqualTo("Title");
        assertThat(loaded.getLocaleValue(Locale.ENGLISH).getDescription()).isEqualTo("Body");

        // 없는 Locale 에 대해서는 기본 Locale 정보를 제공합니다.
        assertThat(loaded.getLocaleValue(Locale.CHINESE).getTitle()).isEqualTo("제목");
        assertThat(loaded.getLocaleValue(Locale.CHINESE).getDescription()).isEqualTo("내용");

        em.remove(loaded);
        em.flush();

        assertThat(em.find(SampleLocaleEntity.class, entity.getId())).isNull();
    }
}
